README.md
# AsyncTCP_SSL Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/AsyncTCP_SSL.svg?)](https://www.ardu-badge.com/AsyncTCP_SSL)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/AsyncTCP_SSL.svg)](https://github.com/khoih-prog/AsyncTCP_SSL/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/AsyncTCP_SSL.svg)](http://github.com/khoih-prog/AsyncTCP_SSL/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of contents

* [Important Note for ESP32_S3](#Important-Note-for-ESP32_S3)
* [Important Change from v1.2.0](#Important-Change-from-v120)
* [Why do we need this AsyncTCP_SSL library](#why-do-we-need-this-AsyncTCP_SSL-library)
  * [Features](#features)
  * [Why Async is better](#why-async-is-better)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Note for Platform IO using ESP32 LittleFS](#note-for-platform-io-using-esp32-littlefs) 
* [HOWTO Fix `Multiple Definitions` Linker Error](#howto-fix-multiple-definitions-linker-error)
* [Note for Platform IO using ESP32 LittleFS](#note-for-platform-io-using-esp32-littlefs)
* [HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)](#howto-use-analogread-with-esp32-running-wifi-andor-bluetooth-btble)
  * [1. ESP32 has 2 ADCs, named ADC1 and ADC2](#1--esp32-has-2-adcs-named-adc1-and-adc2)
  * [2. ESP32 ADCs functions](#2-esp32-adcs-functions)
  * [3. ESP32 WiFi uses ADC2 for WiFi functions](#3-esp32-wifi-uses-adc2-for-wifi-functions)
* [Original documentation](#Original-documentation)
  * [AsyncSSLClient](#AsyncSSLClient)
* [Examples](#examples)
  * [1. multiFileProject](examples/multiFileProject)
* [Debug Terminal Output Samples](#debug-terminal-output-samples) 
  * [1. AsyncHTTPSRequest_ESP on ESP32_DEV](#1-AsyncHTTPSRequest_ESP-on-ESP32_DEV)
  * [2. AsyncHTTPSRequest_ESP on ESP32S2_DEV](#2-AsyncHTTPSRequest_ESP-on-ESP32S2_DEV)
  * [3. AsyncHTTPSRequest_ESP on ESP32C3_DEV](#3-AsyncHTTPSRequest_ESP-on-ESP32C3_DEV)
  * [4. AsyncHTTPSRequest_ESP_WiFiManager on ESP32_DEV](#4-AsyncHTTPSRequest_ESP_WiFiManager-on-ESP32_DEV)
  * [5. AsyncHTTPSRequest_ESP_Multi on ESP32S3_DEV](#5-AsyncHTTPSRequest_ESP_Multi-on-ESP32S3_DEV)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)

---
---

### Important Note for ESP32_S3

**Don't use `ESP32_S3` with core v2.0.4**. Check `already fixed` [ESP32-S3 Powercycling right after uploading a sketch using Arduino IDE and Arduino Core 2.0.4 #7165](https://github.com/espressif/arduino-esp32/issues/7165)

**ESP32_S3 is OK now with core v2.0.5**

---

### Important Change from v1.2.0

Please have a look at [HOWTO Fix `Multiple Definitions` Linker Error](#howto-fix-multiple-definitions-linker-error)

---

### Why do we need this [AsyncTCP_SSL library](https://github.com/khoih-prog/AsyncTCP_SSL)

#### Features

This library is based on, modified from:

1. [Hristo Gochkov's AsyncTCP](https://github.com/me-no-dev/AsyncTCP)
2. [Maarten Fremouw's AsyncTCP](https://github.com/fremouw/AsyncTCP).
3. [Thorsten von Eicken's AsyncTCP](https://github.com/tve/AsyncTCP)

to apply the better and faster **asynchronous** feature of the **powerful** [AsyncTCP Library](https://github.com/me-no-dev/AsyncTCP) with SSL, and will be the base for future and more advanced Async libraries for ESP32, such as AsyncSSLWebServer, AsyncHTTPSRequest, etc.


#### Why Async is better

- Using asynchronous network means that you can handle **more than one connection at the same time**
- **You are called once the request is ready and parsed**
- When you send the response, you are **immediately ready** to handle other connections while the server is taking care of sending the response in the background
- **Speed is OMG**
- **Easy to use API, HTTP Basic and Digest MD5 Authentication (default), ChunkedResponse**
- Easily extensible to handle **any type of content**
- Supports Continue 100
- **Async WebSocket plugin offering different locations without extra servers or ports**
- Async EventSource (Server-Sent Events) plugin to send events to the browser
- URL Rewrite plugin for conditional and permanent url rewrites
- ServeStatic plugin that supports cache, Last-Modified, default index and more
- Simple template processing engine to handle templates


### Currently supported Boards

1. `ESP32` boards, such as ESP32_DEV, etc.
2. `ESP32_S2`-based boards, such as ESP32S2_DEV, ESP32_S2 Saola, etc.
3. `ESP32_C3`-based boards, such as ESP32C3_DEV, etc.
4. `ESP32_S3`-based boards, such as ESP32S3_DEV, etc., using ESP32 core `v2.0.3` or `v2.0.5+`


---
---

## Prerequisites

 1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
 2. [`ESP32 Core 2.0.5+`](https://github.com/espressif/arduino-esp32) for ESP32-based boards. [![Latest release](https://img.shields.io/github/release/espressif/arduino-esp32.svg)](https://github.com/espressif/arduino-esp32/releases/latest/) for ESP32, ESP32_S2, ESP32_C3
 3. [`ESP32 Core 2.0.4`](https://github.com/espressif/arduino-esp32) can't be used for ESP32_S3-based boards. Check `already fixed` [ESP32-S3 Powercycling right after uploading a sketch using Arduino IDE and Arduino Core 2.0.4 #7165](https://github.com/espressif/arduino-esp32/issues/7165)

---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for [**AsyncTCP_SSL**](https://github.com/khoih-prog/AsyncTCP_SSL), then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/AsyncTCP_SSL.svg?)](https://www.ardu-badge.com/AsyncTCP_SSL) for more detailed instructions.

### Manual Install

Another way to install is to:

1. Navigate to [**AsyncTCP_SSL**](https://github.com/khoih-prog/AsyncTCP_SSL) page.
2. Download the latest release `AsyncTCP_SSL-main.zip`.
3. Extract the zip file to `AsyncTCP_SSL-main` directory 
4. Copy whole `AsyncTCP_SSL-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**AsyncTCP_SSL** library](https://registry.platformio.org/libraries/khoih-prog/AsyncTCP_SSL) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/AsyncTCP_SSL/installation). Search for **AsyncTCP_SSL** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)

---
---


### Note for Platform IO using ESP32 LittleFS

In Platform IO, to fix the error when using [`LittleFS_esp32 v1.0`](https://github.com/lorol/LITTLEFS) for ESP32-based boards with ESP32 core v1.0.4- (ESP-IDF v3.2-), uncomment the following line

from

```cpp
//#define CONFIG_LITTLEFS_FOR_IDF_3_2   /* For old IDF - like in release 1.0.4 */
```

to

```cpp
#define CONFIG_LITTLEFS_FOR_IDF_3_2   /* For old IDF - like in release 1.0.4 */
```

It's advisable to use the latest [`LittleFS_esp32 v1.0.5+`](https://github.com/lorol/LITTLEFS) to avoid the issue.

Thanks to [Roshan](https://github.com/solroshan) to report the issue in [Error esp_littlefs.c 'utime_p'](https://github.com/khoih-prog/ESPAsync_WiFiManager/issues/28) 

---
---

### HOWTO Fix `Multiple Definitions` Linker Error

The current library implementation, using `xyz-Impl.h` instead of standard `xyz.cpp`, possibly creates certain `Multiple Definitions` Linker error in certain use cases.

You can include this `.hpp` file

```cpp
// Can be included as many times as necessary, without `Multiple Definitions` Linker Error
#include "AsyncTCP_SSL.hpp"     //https://github.com/khoih-prog/AsyncTCP_SSL
```

in many files. But be sure to use the following `.h` file **in just 1 `.h`, `.cpp` or `.ino` file**, which must **not be included in any other file**, to avoid `Multiple Definitions` Linker Error

```cpp
// To be included only in main(), .ino with setup() to avoid `Multiple Definitions` Linker Error
#include "AsyncTCP_SSL.h"       //https://github.com/khoih-prog/AsyncTCP_SSL
```

Check the new [**multiFileProject** example](examples/multiFileProject) for a `HOWTO` demo.

---
---

### Note for Platform IO using ESP32 LittleFS

In Platform IO, to fix the error when using [`LittleFS_esp32 v1.0`](https://github.com/lorol/LITTLEFS) for ESP32-based boards with ESP32 core v1.0.4- (ESP-IDF v3.2-), uncomment the following line

from

```cpp
//#define CONFIG_LITTLEFS_FOR_IDF_3_2   /* For old IDF - like in release 1.0.4 */
```

to

```cpp
#define CONFIG_LITTLEFS_FOR_IDF_3_2   /* For old IDF - like in release 1.0.4 */
```

It's advisable to use the latest [`LittleFS_esp32 v1.0.5+`](https://github.com/lorol/LITTLEFS) to avoid the issue.

Thanks to [Roshan](https://github.com/solroshan) to report the issue in [Error esp_littlefs.c 'utime_p'](https://github.com/khoih-prog/ESPAsync_WiFiManager/issues/28) 

---
---

### HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)

Please have a look at [**ESP_WiFiManager Issue 39: Not able to read analog port when using the autoconnect example**](https://github.com/khoih-prog/ESP_WiFiManager/issues/39) to have more detailed description and solution of the issue.

#### 1.  ESP32 has 2 ADCs, named ADC1 and ADC2

#### 2. ESP32 ADCs functions

- ADC1 controls ADC function for pins **GPIO32-GPIO39**
- ADC2 controls ADC function for pins **GPIO0, 2, 4, 12-15, 25-27**

#### 3.. ESP32 WiFi uses ADC2 for WiFi functions

Look in file [**adc_common.c**](https://github.com/espressif/esp-idf/blob/master/components/driver/adc_common.c#L61)

> In ADC2, there're two locks used for different cases:
> 1. lock shared with app and Wi-Fi:
>    ESP32:
>         When Wi-Fi using the ADC2, we assume it will never stop, so app checks the lock and returns immediately if failed.
>    ESP32S2:
>         The controller's control over the ADC is determined by the arbiter. There is no need to control by lock.
> 
> 2. lock shared between tasks:
>    when several tasks sharing the ADC2, we want to guarantee
>    all the requests will be handled.
>    Since conversions are short (about 31us), app returns the lock very soon,
>    we use a spinlock to stand there waiting to do conversions one by one.
> 
> adc2_spinlock should be acquired first, then adc2_wifi_lock or rtc_spinlock.


- In order to use ADC2 for other functions, we have to **acquire complicated firmware locks and very difficult to do**
- So, it's not advisable to use ADC2 with WiFi/BlueTooth (BT/BLE).
- Use ADC1, and pins GPIO32-GPIO39
- If somehow it's a must to use those pins serviced by ADC2 (**GPIO0, 2, 4, 12, 13, 14, 15, 25, 26 and 27**), use the **fix mentioned at the end** of [**ESP_WiFiManager Issue 39: Not able to read analog port when using the autoconnect example**](https://github.com/khoih-prog/ESP_WiFiManager/issues/39) to work with ESP32 WiFi/BlueTooth (BT/BLE).

---
---


## Original documentation

For ESP32, check [AsyncTCP Library](https://github.com/me-no-dev/AsyncTCP)

This is a fully asynchronous SSL TCP library, aimed at enabling trouble-free, multi-connection network environment for Espressif's ESP32 MCUs.

### AsyncSSLClient

The base classes on which everything else is built. They expose all possible scenarios, but are really raw and require more skills to use.

---
---

### Examples

 1. [multiFileProject](examples/multiFileProject). **New**

---
---

### Debug Terminal Output Samples

#### 1. AsyncHTTPSRequest_ESP on ESP32_DEV

Following is the debug terminal when running example [AsyncHTTPSRequest_ESP](https://github.com/khoih-prog/AsyncHTTPSRequest_Generic/tree/main/examples/AsyncHTTPSRequest_ESP) on ESP32_DEV to demonstrate the operation of SSL Async HTTPS request, using [AsyncTCP_SSL Library](https://github.com/khoih-prog/AsyncTCP_SSL).

```cpp
Starting AsyncHTTPSRequest_ESP using ESP32_DEV
AsyncTCP_SSL v1.3.1
AsyncHTTPSRequest_Generic v2.1.2
Connecting to WiFi SSID: HueNet1
.......
AsyncHTTPSRequest @ IP : 192.168.2.133

**************************************
abbreviation: EDT
client_ip: aaa.bbb.ccc.ddd
datetime: 2022-09-18T01:35:32.451975-04:00
day_of_week: 0
day_of_year: 261
dst: true
dst_from: 2022-03-13T07:00:00+00:00
dst_offset: 3600
dst_until: 2022-11-06T06:00:00+00:00
raw_offset: -18000
timezone: America/Toronto
unixtime: 1663479332
utc_datetime: 2022-09-18T05:35:32.451975+00:00
utc_offset: -04:00
week_number: 37
**************************************
HHHHHH
**************************************
abbreviation: EDT
client_ip: aaa.bbb.ccc.ddd
datetime: 2022-09-18T01:36:31.992595-04:00
day_of_week: 0
day_of_year: 261
dst: true
dst_from: 2022-03-13T07:00:00+00:00
dst_offset: 3600
dst_until: 2022-11-06T06:00:00+00:00
raw_offset: -18000
timezone: America/Toronto
unixtime: 1663479391
utc_datetime: 2022-09-18T05:36:31.992595+00:00
utc_offset: -04:00
week_number: 37
```
---

#### 2. AsyncHTTPSRequest_ESP on ESP32S2_DEV


changelog.md
# AsyncTCP_SSL Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/AsyncTCP_SSL.svg?)](https://www.ardu-badge.com/AsyncTCP_SSL)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/AsyncTCP_SSL.svg)](https://github.com/khoih-prog/AsyncTCP_SSL/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/AsyncTCP_SSL.svg)](http://github.com/khoih-prog/AsyncTCP_SSL/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v1.3.1](#Releases-v131)
  * [Releases v1.3.0](#Releases-v130)
  * [Releases v1.2.0](#Releases-v120)
  * [Releases v1.1.0](#Releases-v110)
  * [Initial Releases v1.0.0](#Initial-Releases-v100)

---
---

## Changelog

### Releases v1.3.1

1. Increase `ASYNC_QUEUE_LENGTH` to default 512 from 32 and make it user-configurable
2. Increase `ASYNC_TCP_PRIORITY` to default 10 from 3, and make it user-configurable


### Releases v1.3.0

1. Remove hard-code if possible
2. Improve debug messages by adding functions to display `error/state messages` instead of `cryptic error/state number`
3. Clean up
4. Add support to `ESP32_S3`, using ESP32 core `v2.0.3`. **Don't use `ESP32_S3` with core v2.0.4**. Check [ESP32-S3 Powercycling right after uploading a sketch using Arduino IDE and Arduino Core 2.0.4 #7165](https://github.com/espressif/arduino-esp32/issues/7165)

### Releases v1.2.0

1. Fix `multiple-definitions` linker error. Drop `src_cpp` and `src_h` directories
2. Add example [multiFileProject](examples/multiFileProject) to demo for multiple-file project.
3. Add `platformio.ini`

### Releases v1.1.0

1. Fix duplication bug when using `src_h`
2. Enable coexistence with AsyncTCP

### Initial Releases v1.0.0

1. Initial coding to support **ESP32**





src/AsyncTCP_SSL.hpp
/****************************************************************************************************************************
  AsyncTCP_SSL.hpp
   
  AsyncTCP_SSL is a library for ESP32
  
  Based on and modified from :
  
  1) AsyncTCP (https://github.com/me-no-dev/ESPAsyncTCP)
  2) AsyncTCP (https://github.com/tve/AsyncTCP)
  
  Built by Khoi Hoang https://github.com/khoih-prog/AsyncTCP_SSL
  
  This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License 
  as published bythe Free Software Foundation, either version 3 of the License, or (at your option) any later version.
  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 
  Version: 1.3.1
  
  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0    K Hoang     21/10/2021 Initial coding to support only ESP32
  1.1.0    K Hoang     22/10/2021 Fix bug. Enable coexistence with AsyncTCP
  1.2.0    K Hoang     23/01/2022 Fix `multiple-definitions` linker error
  1.3.0    K Hoang     04/09/2022 Clean up. Remove hard-code if possible
  1.3.1    K Hoang     18/09/2022 Improve stability. Make queue length user-configurable
 *****************************************************************************************************************************/
 
/*
  Asynchronous TCP library for Espressif MCUs

  Copyright (c) 2016 Hristo Gochkov. All rights reserved.
  This file is part of the esp8266 core for Arduino environment.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef ASYNCTCP_SSL_HPP
#define ASYNCTCP_SSL_HPP

#if !defined(ESP32)
  #error This AsyncTCP_SSL library is supporting only ESP32
#endif

/////////////////////////////////////////////////

#define ASYNC_TCP_SSL_VERSION             "AsyncTCP_SSL v1.3.1"

#define ASYNC_TCP_SSL_VERSION_MAJOR       1
#define ASYNC_TCP_SSL_VERSION_MINOR       3
#define ASYNC_TCP_SSL_VERSION_PATCH       1

#define ASYNC_TCP_SSL_VERSION_INT         1003001

/////////////////////////////////////////////////

#include "Arduino.h"

#include "IPAddress.h"
#include "sdkconfig.h"
#include <functional>
#include <string>

// in ./libraries/WiFiClientSecure/src/ssl_client.h for ESP32
#include <ssl_client.h>

#include "tcp_mbedtls.h"

extern "C" 
{
  #include "freertos/FreeRTOS.h"
  #include "freertos/semphr.h"
  #include "lwip/pbuf.h"
}

#include "AsyncTCP_SSL_Debug.h"

/*****************************************************
// Defined in tools/sdk/esp32s3/include/lwip/lwip/src/include/lwip/tcpbase.h
enum tcp_state 
{
  CLOSED      = 0,
  LISTEN      = 1,
  SYN_SENT    = 2,
  SYN_RCVD    = 3,
  ESTABLISHED = 4,
  FIN_WAIT_1  = 5,
  FIN_WAIT_2  = 6,
  CLOSE_WAIT  = 7,
  CLOSING     = 8,
  LAST_ACK    = 9,
  TIME_WAIT   = 10
};
*****************************************************/

#define INVALID_CLOSED_SLOT        -1

//////////////////////////////////////////////////////////////////////////////////////////

//If core is not defined, then we are running in Arduino or PIO
#ifndef CONFIG_ASYNC_TCP_RUNNING_CORE
  #define CONFIG_ASYNC_TCP_RUNNING_CORE     		-1    //any available core
  #define CONFIG_ASYNC_TCP_USE_WDT          		1     //if enabled, adds between 33us and 200us per event
#endif

/////////////////////////////////////////////////

// Make ASYNC_QUEUE_LENGTH adjustable in sketch. Default 512
#ifndef ASYNC_QUEUE_LENGTH
  #define ASYNC_QUEUE_LENGTH 		512
#endif

// Make ASYNC_TCP_PRIORITY user-adjustable in sketch. Default 10, can't be less than 4
#if !defined(CONFIG_ASYNC_TCP_PRIORITY)
  #define CONFIG_ASYNC_TCP_PRIORITY 	(10)
#elif (CONFIG_ASYNC_TCP_PRIORITY < 4)
  #undef CONFIG_ASYNC_TCP_PRIORITY
  #define CONFIG_ASYNC_TCP_PRIORITY 	(4)
  #warning Adjust CONFIG_ASYNC_TCP_PRIORITY to 4
#endif

/////////////////////////////////////////////////

#define ASYNC_MAX_ACK_TIME      5000
#define ASYNC_WRITE_FLAG_COPY   0x01    //will allocate new buffer to hold the data while sending (else will hold reference to the data given)
#define ASYNC_WRITE_FLAG_MORE   0x02    //will not send PSH flag, meaning that there should be more data to be sent before the application should react.
#define SSL_HANDSHAKE_TIMEOUT   5000    // timeout to complete SSL handshake

//////////////////////////////////////////////////////////////////////////////////////////////

class AsyncSSLClient;

typedef std::function<void(void*, AsyncSSLClient*)> AcConnectHandlerSSL;
typedef std::function<void(void*, AsyncSSLClient*, size_t len, uint32_t time)> AcAckHandlerSSL;
typedef std::function<void(void*, AsyncSSLClient*, int8_t error)> AcErrorHandlerSSL;
typedef std::function<void(void*, AsyncSSLClient*, void *data, size_t len)> AcDataHandlerSSL;
typedef std::function<void(void*, AsyncSSLClient*, struct pbuf *pb)> AcPacketHandlerSSL;
typedef std::function<void(void*, AsyncSSLClient*, uint32_t time)> AcTimeoutHandlerSSL;

/////////////////////////////////////////////////

struct tcp_pcb;
struct ip_addr;

//////////////////////////////////////////////////////////////////////////////////////////////

class AsyncSSLClient 
{
  public:
    AsyncSSLClient(tcp_pcb* pcb = 0);
    virtual ~AsyncSSLClient();

    AsyncSSLClient & operator=(const AsyncSSLClient &other);
    AsyncSSLClient & operator+=(const AsyncSSLClient &other);

    bool operator==(const AsyncSSLClient &other);

    bool operator!=(const AsyncSSLClient &other) 
    {
      return !(*this == other);
    }

    bool    connect(IPAddress ip, uint16_t port, bool secure = false);
    bool    connect(const char* host, uint16_t port,  bool secure = false);
    void    setRootCa(const char* rootca, const size_t len);
    void    setClientCert(const char* cli_cert, const size_t len);
    void    setClientKey(const char* cli_key, const size_t len);
    void    setPsk(const char* psk_ident, const char* psk);

    void    close(bool now = false);
    void    stop();
    int8_t  abort();
    bool    free();

    bool    canSend();        //ack is not pending
    size_t  add(const char* data, size_t size, uint8_t apiflags = ASYNC_WRITE_FLAG_COPY); //add for sending
    bool    send();           //send all data added with the method above

    virtual size_t  space();  //space available in the TCP window
    //write equals add()+send()
    virtual size_t  write(const char* data);
    virtual size_t  write(const char* data, size_t size, uint8_t apiflags = ASYNC_WRITE_FLAG_COPY); //only when canSend() == true

    uint8_t state();
    bool    connecting();
    bool    connected();
    bool    disconnecting();
    bool    disconnected();
    bool    freeable();       //disconnected or disconnecting

    uint16_t  getMss();

    uint32_t  getRxTimeout();
    void      setRxTimeout(uint32_t timeout);     //no RX data timeout for the connection in seconds

    uint32_t  getAckTimeout();
    void      setAckTimeout(uint32_t timeout);    //no ACK timeout for the last sent packet in milliseconds

    void      setNoDelay(bool nodelay);
    bool      getNoDelay();

    uint32_t  getRemoteAddress();
    uint16_t  getRemotePort();
    uint32_t  getLocalAddress();
    uint16_t  getLocalPort();

    //compatibility
    IPAddress remoteIP();
    uint16_t  remotePort();
    IPAddress localIP();
    uint16_t  localPort();

    void    onConnect(AcConnectHandlerSSL cb, void* arg = 0);      //on successful connect
    void    onDisconnect(AcConnectHandlerSSL cb, void* arg = 0);   //disconnected
    void    onAck(AcAckHandlerSSL cb, void* arg = 0);              //ack received
    void    onError(AcErrorHandlerSSL cb, void* arg = 0);          //unsuccessful connect or error
    void    onData(AcDataHandlerSSL cb, void* arg = 0);            //data received (called if onPacket is not used)
    void    onPacket(AcPacketHandlerSSL cb, void* arg = 0);        //data received
    void    onTimeout(AcTimeoutHandlerSSL cb, void* arg = 0);      //ack timeout
    void    onPoll(AcConnectHandlerSSL cb, void* arg = 0);         //every 125ms when connected

    void    ackPacket(struct pbuf * pb);//ack pbuf from onPacket
    size_t  ack(size_t len); //ack data that you have not acked using the method below
    
    void ackLater() 
    {
      _ack_pcb = false;  //will not ack the current packet. Call from onData
    }

    const char *  errorToString(int8_t error);
    const char *  stateToString();

    //Do not use any of the functions below!
    static int8_t _s_poll(void *arg, struct tcp_pcb *tpcb);
    static int8_t _s_recv(void *arg, struct tcp_pcb *tpcb, struct pbuf *pb, int8_t err);
    static int8_t _s_fin(void *arg, struct tcp_pcb *tpcb, int8_t err);
    static int8_t _s_lwip_fin(void *arg, struct tcp_pcb *tpcb, int8_t err);
    static void   _s_error(void *arg, int8_t err);
    static int8_t _s_sent(void *arg, struct tcp_pcb *tpcb, uint16_t len);
    static int8_t _s_connected(void* arg, void* tpcb, int8_t err);
    static void   _s_dns_found(const char *name, struct ip_addr *ipaddr, void *arg);
    
    static void _s_data(void *arg, struct tcp_pcb *tcp, uint8_t * data, size_t len);
    static void _s_handshake(void *arg, struct tcp_pcb *tcp, struct tcp_ssl_pcb* ssl);
    static void _s_ssl_error(void *arg, struct tcp_pcb *tcp, int8_t err);

    int8_t      _recv(tcp_pcb* pcb, pbuf* pb, int8_t err);
    
    tcp_pcb * pcb() 
    {
      return _pcb;
    }

    //KH
    int8_t getClosed_Slot() 
    {
      return _closed_slot;
    }
    //////

  protected:
    tcp_pcb*        _pcb;
    std::string     _hostname;
    int8_t          _closed_slot;

    AcConnectHandlerSSL   _connect_cb;
    void*                 _connect_cb_arg;
    AcConnectHandlerSSL   _discard_cb;
    void*                 _discard_cb_arg;
    AcAckHandlerSSL       _sent_cb;
    void*                 _sent_cb_arg;
    AcErrorHandlerSSL     _error_cb;
    void*                 _error_cb_arg;
    AcDataHandlerSSL      _recv_cb;
    void*                 _recv_cb_arg;
    AcPacketHandlerSSL    _pb_cb;
    void*                 _pb_cb_arg;
    AcTimeoutHandlerSSL   _timeout_cb;
    void*                 _timeout_cb_arg;
    AcConnectHandlerSSL   _poll_cb;
    void*                 _poll_cb_arg;

    bool      _pcb_busy;
    uint32_t  _pcb_sent_at;
    bool      _ack_pcb;
    uint32_t  _rx_ack_len;
    uint32_t  _rx_last_packet;
    uint32_t  _rx_since_timeout;
    uint32_t  _ack_timeout;
    uint16_t  _connect_port;

    ////// SSL
    size_t  _root_ca_len;
    char*   _root_ca;
    size_t  _cli_cert_len;
    char*   _cli_cert;
    size_t  _cli_key_len;
    char*   _cli_key;
    bool    _pcb_secure;
    bool    _handshake_done;
    
    const char* _psk_ident;
    const char* _psk;
    //////

    int8_t  _close();
    void    _free_closed_slot();
    void    _allocate_closed_slot();
    int8_t  _connected(void* pcb, int8_t err);
    void    _error(int8_t err);
    int8_t  _poll(tcp_pcb* pcb);
    int8_t  _sent(tcp_pcb* pcb, uint16_t len);
    int8_t  _fin(tcp_pcb* pcb, int8_t err);
    int8_t  _lwip_fin(tcp_pcb* pcb, int8_t err);
    void    _dns_found(struct ip_addr *ipaddr);
    
    ////// SSL
    void    _ssl_error(int8_t err);
    //////

  public:
    AsyncSSLClient* prev;
    AsyncSSLClient* next;
};

typedef std::function<int(void* arg, const char *filename, uint8_t **buf)> AcSSlFileHandlerSSL;

//////////////////////////////////////////////////////////////////////////////////////////////

class AsyncSSLServer 
{
  public:
    AsyncSSLServer(IPAddress addr, uint16_t port);
    AsyncSSLServer(uint16_t port);
    ~AsyncSSLServer();
    
    void onClient(AcConnectHandlerSSL cb, void* arg);
    
    // Dummy, so it compiles with ESP Async WebServer library enabled.
    void    onSslFileRequest(AcSSlFileHandlerSSL cb, void* arg) {};
    void    beginSecure(const char *cert, const char *private_key_file, const char *password) {};

    void    begin();
    void    end();
    void    setNoDelay(bool nodelay);
    bool    getNoDelay();
    uint8_t status();

    //Do not use any of the functions below!
    static int8_t _s_accept(void *arg, tcp_pcb* newpcb, int8_t err);
    static int8_t _s_accepted(void *arg, AsyncSSLClient* client);

  protected:
    uint16_t  _port;
    IPAddress _addr;
    bool      _noDelay;
    
    tcp_pcb*              _pcb;
    
    AcConnectHandlerSSL   _connect_cb;   
    void*                 _connect_cb_arg;

    int8_t _accept(tcp_pcb* newpcb, int8_t err);
    int8_t _accepted(AsyncSSLClient* client);
};

//////////////////////////////////////////////////////////////////////////////////////////////

#endif /* ASYNCTCP_SSL_HPP */
